Add SummaryDocumentActivity creation

dwhenry opened the pull request:
Create a new TelephoneAppointment::SummaryDocumentActivity
with appointment_id and guider_uid (from signon).

Calling save on the object will post to 
`/api/v1/summary_documents`, return true is the
save was successful and false otherwise. In addition
the validation errors will be accessible via the `error`
method

dwhenry committed:
Add SummaryDocumentActivity creation

Create a new TelephoneAppointment::SummaryDocumentActivity
with appointment_id and guider_uid (from signon).

Calling save on the object will post to
`/api/v1/summary_documents`, return true is the
save was successful and false otherwise. In addition
the validation errors will be accessible via the `error`
method

dwhenry committed:
Remove rubocop BlockLength check from spec folder

benlovell commented on README.md:
````diff
@@ -22,7 +22,7 @@ Or install it yourself as:
 ```ruby
 
-TelephoneAppointments::SummaryDocument.create(params) # returns a Summary Document
+TelephoneAppointments::SummaryDocument.create(appointment_id: id, guider_uid: uid) # returns a Summary Document
````
This returns a boolean, not a summary document.

benlovell commented on lib/telephone_appointments/response.rb:
```diff
@@ -0,0 +1,17 @@
+module TelephoneAppointments
+  class Response
+    def initialize(response)
+      @response = response
```
Can we add a private reader for accessing this elsewhere internally to this class?

benlovell commented on lib/telephone_appointments/response.rb:
```diff
@@ -0,0 +1,17 @@
+
+    def errors
+      return {} if success?
+      return {} if @response.body.nil? || @response.body == ''
```
Why not gather these two lines up?

benlovell commented on lib/telephone_appointments/stub_api.rb:
```diff
@@ -1,4 +1,7 @@
 module TelephoneAppointments
   class StubApi
+    def post(_path, _form_data)
```
I'd just favour `post(*)` as the arguments are inconsequential really and the arity isn't worth preserving.

benlovell commented on lib/telephone_appointments/summary_document_activity.rb:
```diff
@@ -0,0 +1,24 @@
+
+    def save
+      @response = TelephoneAppointments.api.post(
+        "/api/#{API_VERSION}/#{PATH}",
```
Technically speaking, all of this is the path. Is there much use in calling out the `API_VERSION` and `PATH` here? Instead of just putting the whole lot in a string without interpolation I mean.

benlovell commented on lib/telephone_appointments/summary_document_activity.rb:
```diff
@@ -0,0 +1,24 @@
+    end
+
+    def errors
+      @response.errors
```
What happens if I call this method without calling `save` first? We should guard against that for a consistent API.

benlovell commented on spec/features/create_a_summary_document_activity_spec.rb:
```diff
@@ -0,0 +1,24 @@
+    expect(summary_document.save).to be false
+    expect(summary_document.errors).to eq(
+      'appointment' => ['can\'t be blank'],
+      'owner' => ['can\'t be blank']
```
The response keys for the errors don't match the given arguments so it may be difficult for consumers to tie the two up. This is more of an observation than a hard requirement though, so feel free to pass this over.

benlovell commented on spec/telephone_appointments/response_spec.rb:
```diff
@@ -0,0 +1,41 @@
+      expect(subject).not_to be_success
+    end
+
+    it 'does not have any errors' do
```
I think you've copy pasted this spec description as it doesn't match with what's actually happening here.

benlovell commented on spec/telephone_appointments/response_spec.rb:
```diff
@@ -0,0 +1,41 @@
+      expect(subject.errors).to eq('trade' => 'invalid')
+    end
+
+    context 'and no body is returned' do
```
This should be `when no body is returned`.

benlovell commented on spec/telephone_appointments/response_spec.rb:
```diff
@@ -0,0 +1,41 @@
+  context 'when the response is valid' do
+    let(:response) { Net::HTTPSuccess.new(nil, nil, nil) }
+
+    it 'should be a success' do
```
Use the unequivocal style for spec descriptions. `is a success` reads better here. Same elsewhere in this spec file.

benlovell commented on telephone_appointments.gemspec:
```diff
@@ -26,4 +26,6 @@ Gem::Specification.new do |spec|
   spec.add_development_dependency 'rspec', '~> 3.0'
   spec.add_development_dependency 'rubocop', '~> 0.41'
+  spec.add_development_dependency 'vcr', '~> 3.0'
+  spec.add_development_dependency 'webmock', '~> 2.1'
```
It's not in this diff but you don't need to reference `GlobalID`. Also these are pessimistically pinned at old versions. We should get them up to date as far as possible first.

benlovell reviewed (changes requested):
You might want to grep across the project for `locations` as there are a few occurrences that should be squashed / replaced.

dwhenry commented:
@benlovell I think I have addressed all your points. 

I have renamed the `guider_uid` parameter to `owner_uid` as I think this is the best solution for debugging on the error messages.

benlovell commented on spec/telephone_appointments/summary_document_activity_spec.rb:
```diff
@@ -0,0 +1,31 @@
+RSpec.describe TelephoneAppointments::SummaryDocumentActivity do
+  subject { described_class.new(appointment_id: 12, owner_uid: 'abc') }
+
+  describe '.errors' do
```
Sorry I missed this - the correct notation here is `#errors`.

benlovell reviewed (approved):
Just that notation thing but other than that 🚢
